src/ILVerify/src/Verifier.cs

```diff
@@ -0,0 +1,224 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using System.Resources;
+using System.Text;
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILVerify
+{
+    public delegate bool ShouldVerifyMethod(string name);
+
+    public interface IResolver
+    {
+        PEReader Resolve(AssemblyName name);
+    }
+
+    public class Verifier
+    {
+        private Lazy<ResourceManager> _stringResourceManager =
+            new Lazy<ResourceManager>(() => new ResourceManager("ILVerify.Resources.Strings", Assembly.GetExecutingAssembly()));
+
+        private SimpleTypeSystemContext _typeSystemContext;
+
+        public ShouldVerifyMethod ShouldVerifyMethod { private get; set; }
+
+        public Verifier(IResolver resolver)
+        {
+            _typeSystemContext = new SimpleTypeSystemContext(resolver);
+        }
+
+        internal Verifier(SimpleTypeSystemContext context)
+        {
+            _typeSystemContext = context;
+        }
+
+        public void SetSystemModuleName(AssemblyName name)
+        {
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(name));
+        }
+
+        public VerificationResult Verify(AssemblyName moduleToVerify)
+        {
+            if (moduleToVerify == null)
+            {
+                throw new ArgumentNullException(nameof(moduleToVerify));
+            }
+
+            try
+            {
+                if (_typeSystemContext.SystemModule is null)
+                {
+                    return new VerificationResult() { NumErrors = 1, Message = "No system module specified" };
+                }
+
+                EcmaModule module = _typeSystemContext.GetModule(moduleToVerify);
+                return VerifyModule(module, moduleToVerify.Name);
+            }
+            catch (VerifierException e)
+            {
+                return new VerificationResult() { NumErrors = 1, Message = e.Message };
+            }
+        }
+
+        private VerificationResult VerifyModule(EcmaModule module, string path)
+        {
+            foreach (var methodHandle in module.MetadataReader.MethodDefinitions)
+            {
+                var method = (EcmaMethod)module.GetMethod(methodHandle);
+
+                var methodIL = EcmaMethodIL.Create(method);
+                if (methodIL == null)
+                {
+                    continue;
+                }
+
+                var methodName = method.ToString();
```

Comment:
`ToString()` on the type system objects is a debugging helper, not an API. We reserve the right to change the output any time to aid debuggability. I recently overhauled it in #5065.